scalardb-test/src/main/java/kelpie/scalardb/transfer/TransferWithTwoPhaseCommitTransactionProcessorWithRetry.java

```diff
@@ -0,0 +1,115 @@
+  private final int numAccounts;
+  private final Retry retry;
+  private final LongAdder retryCount = new LongAdder();
+
+  public TransferWithTwoPhaseCommitTransactionProcessorWithRetry(Config config) {
+    super(config);
+    manager1 = TransferCommon.getTwoPhaseCommitTransactionManager1(config);
+    manager2 = TransferCommon.getTwoPhaseCommitTransactionManager2(config);
+
+    numAccounts = (int) config.getUserLong("test_config", "num_accounts");
+
+    RetryConfig retryConfig =
+        RetryConfig.custom()
+            .maxAttempts(MAX_RETRIES)
+            .intervalFunction(
+                IntervalFunction.ofExponentialBackoff(INITIAL_INTERNAL_MILLS, MULTIPLIER))
+            .retryExceptions(CrudConflictException.class, CommitConflictException.class)
+            .build();
+    retry = Retry.of("TransferWithTwoPhaseCommitTransactionProcessorWithRetry", retryConfig);
+    retry
+        .getEventPublisher()
+        .onRetry(
+            e -> {
+              logWarn(e.toString(), e.getLastThrowable());
+              retryCount.increment();
+            });
+  }
+
+  @Override
+  public void executeEach() throws Exception {
+    int fromId = ThreadLocalRandom.current().nextInt(numAccounts);
+    int toId = ThreadLocalRandom.current().nextInt(numAccounts);
+    int amount = ThreadLocalRandom.current().nextInt(1000) + 1;
+
+    try {
+      Retry.decorateCheckedRunnable(retry, () -> transfer(fromId, toId, amount)).run();
+    } catch (Throwable t) {
+      if (t instanceof Exception) {
+        throw (Exception) t;
+      }
+
+      throw new Exception(t);
+    }
+  }
+
+  @Override
+  public void close() {
+    manager1.close();
+    manager2.close();
+
+    setState(Json.createObjectBuilder().add("retry_count", retryCount.longValue()).build());
+  }
+
+  private void transfer(int fromId, int toId, int amount) throws Exception {
+    int fromType = 0;
+    int toType = 0;
+    if (fromId == toId) {
+      toType = 1; // transfer between the same account
+    }
+
+    TwoPhaseCommitTransaction tx1 = manager1.start();
+    TwoPhaseCommitTransaction tx2 = manager2.join(tx1.getId());
+    try {
+      Get fromGet = TransferCommon.prepareGet(fromId, fromType);
+      Optional<Result> fromResult = tx1.get(fromGet);
+      int fromBalance = TransferCommon.getBalanceFromResult(fromResult.get());
+      Put fromPut = TransferCommon.preparePut(fromId, fromType, fromBalance - amount);
+      tx1.put(fromPut);
+
+      Get toGet = TransferCommon.prepareGet(toId, toType);
+      Optional<Result> toResult = tx2.get(toGet);
+      int toBalance = TransferCommon.getBalanceFromResult(toResult.get());
+      Put toPut = TransferCommon.preparePut(toId, toType, toBalance + amount);
+      tx2.put(toPut);
+
+      tx1.prepare();
+      tx2.prepare();
+      tx1.validate();
+      tx2.validate();
+      tx1.commit();
+      tx2.commit();
+    } catch (Exception e) {
+      tx1.rollback();
+      tx2.rollback();
```

Comment:
I first thought any exception from `tx1.rollback` should be caught and `tx2.rollback` should be always executed. But it's not always the best way as it could result in a partial rollback. So, on second thought, I think we can leave this rollbacks as-is. What do you think?

Reply:
Ah, thank you for pointing it out! I think we should catch the exceptions from the rollback. I will fix it. Thanks.

Reply:
Fixed in ee55fbd47e00efbfe1ce9b8401d1d19459d31387.